File: boj/graph/bfs/14442/14442.py
from collections import deque
import sys
input = sys.stdin.readline

# 입력 처리
n, m, k = map(int, input().strip().split())
map_input = [list(map(int, input().strip())) for _ in range(n)]

def bfs():
    dr = [0,0,1,-1]
    dc = [1,-1,0,0]
    queue = deque([(0,0,0)])
    check = [[[1000001] * (k+1) for _ in range(m)] for _ in range(n)]
    check[0][0][0] = 1
    while queue:
        r, c, break_cnt = queue.popleft()
        
        if r == n - 1 and c == m - 1:
            return check[r][c][break_cnt]
        for i in range(4):
            nr = r + dr[i]
            nc = c + dc[i]
            if not (0 <= nr < n and 0 <= nc < m):
                continue
            #벽을 부수지 않고 이동가능한 경우
            if map_input[nr][nc] == 0 and check[nr][nc][break_cnt] > check[r][c][break_cnt] + 1:
                check[nr][nc][break_cnt] = check[r][c][break_cnt] + 1
                queue.append((nr,nc,break_cnt))
            if map_input[nr][nc] == 1 and break_cnt < k and check[nr][nc][break_cnt + 1] > check[r][c][break_cnt] + 1:
                check[nr][nc][break_cnt + 1] = check[r][c][break_cnt] + 1
                queue.append((nr,nc,break_cnt+1))
    return -1

print(bfs())

File: boj/graph/bfs/14442/14442.cpp
#include <iostream>
#include <bits/stdc++.h>
using namespace std;

int n, m, k;
vector<vector<int>> map_input;

int dr[4] = {0, 0, 1, -1};
int dc[4] = {1, -1, 0, 0};

int bfs() {
  queue<tuple<int, int, int>> q;
  q.push({0, 0, 0});
  
  vector<vector<vector<int>>> check(n, vector<vector<int>>(m, vector<int>(k + 1, 0)));
  check[0][0][0] = 1;
  
  while (!q.empty()) {
    auto [r, c, break_cnt] = q.front();
    q.pop();
    
    if (r == n - 1 && c == m - 1) {
      return check[r][c][break_cnt];
    }
    
    for (int i = 0; i < 4; i++) {
      int nr = r + dr[i];
      int nc = c + dc[i];
      
      if (nr < 0 || nr >= n || nc < 0 || nc >= m) {
        continue;
      }
      
      if (map_input[nr][nc] == 0 && check[nr][nc][break_cnt] == 0) {
        check[nr][nc][break_cnt] = check[r][c][break_cnt] + 1;
        q.push({nr, nc, break_cnt});
      }
      
      if (map_input[nr][nc] == 1 && break_cnt < k && check[nr][nc][break_cnt + 1] == 0) {
        check[nr][nc][break_cnt + 1] = check[r][c][break_cnt] + 1;
        q.push({nr, nc, break_cnt + 1});
      }
    }
  }
  
  return -1;
}

int main() {
  ios::sync_with_stdio(false);
  cin.tie(nullptr);
  
  cin >> n >> m >> k;
  map_input.resize(n, vector<int>(m));
  
  for (int i = 0; i < n; i++) {
    string line;
    cin >> line;
    for (int j = 0; j < m; j++) {
      map_input[i][j] = line[j] - '0';
    }
  }
  
  cout << bfs() << "\n";
  return 0;
}
